### ISSUE TITLE
Phantom container prevents docker from creating new containers with the same name

### ISSUE THREAD

**Issue by yunjing** [moby/moby#21198]
We noticed it after switching to `overlay` storage. The problem is that we regularly destroying containers and starting new ones with the same name. However, sometimes docker fails to destroy the old containers cleanly, and that prevents us from creating new containers with the same name. Specifically, we noticed the following logs when creating a new container:

```
Mar 14 23:26:15 localhost docker[20200]: time="2016-03-14T23:26:15.285651105Z" level=error msg="Handler for GET /v1.22/containers/multi15/json returned error: No such container: multi15"
Mar 14 23:26:15 localhost docker[20200]: time="2016-03-14T23:26:15.286292324Z" level=error msg="Handler for GET /v1.22/images/multi15/json returned error: No such image: multi15"
Mar 14 23:26:15 localhost docker[20200]: time="2016-03-14T23:26:15.325107592Z" level=error msg="Handler for POST /v1.22/containers/create returned error: Conflict. The name \"/multi15\" is already in use by container 0c8a8c930b236024ab9497bf5ff7e35736daab98bc0edf25ebabb35cbf913435. You have to remove (or rename) that container to be able to reuse that name."
```

Docker reports that `multi15` does not exist but it also claims to have naming conflict. Using docker command, we cannot list the container either by name (`multi15`) or by id (`0c8a8c930b236024ab9497bf5ff7e35736daab98bc0edf25ebabb35cbf913435`). This issue persists until we restart docker daemon.

**Output of `docker version`:**

```
Client:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 21:40:35 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 21:40:35 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 30
 Running: 30
 Paused: 0
 Stopped: 0
Images: 3
Server Version: 1.10.2
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Plugins:
 Volume: local
 Network: bridge null host
Kernel Version: 4.2.0-25-generic
Operating System: Ubuntu 15.10
OSType: linux
Architecture: x86_64
CPUs: 32
Total Memory: 28.27 GiB
Name: app-multi2
ID: LIUR:TIO3:RRZQ:HDNL:4ZSL:XPAQ:QHGC:2ND6:35LB:JNXK:PUEZ:QQ4Z
WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
We are running docker on google cloud VMs.

**Steps to reproduce the issue:**
1. Delete a container
2. Pull a new image
3. Start a new container with the new image with the same name as the deleted container
Note that this seems to be a race condition, and it does not always happens. It also seems to happen more often when we delete/start multiple containers at the same time.

**Describe the results you received:**
Cannot create a container with certain name while the old container with the same name should have been deleted.

**Describe the results you expected:**
Successfully create a container with certain name while the old container with the same name should have been deleted.

**Additional information you deem important (e.g. issue happens only occasionally):**
After restarting docker daemon, we can successfully start the container with the same name.

**Comment by cpuguy83** [comment]
Thanks, I'll look into this.

**Comment by yunjing** [comment]
Thanks. Empirically, it happens to us about once every 300 container creation.

**Comment by yunjing** [comment]
@cpuguy83 I wonder if there is any update on this? I want to add that we are using `MountFlags=private` for those VMs.

**Comment by tnucera** [comment]
Same issue

> Server Version: 1.10.3
Storage Driver: overlay

**Comment by IIIEII** [comment]
I have similar situation in such scenario:
1. Start docker-volume-netshare (cifs) plugin
2. Create new volume with cifs driver and some options (username/password) - in plugin log I can see that Create is called with my options
3. Use volume with new container - works correctly
4. Restart plugin
5. `docker volume ls` shows no volumes for cifs driver
6. Create volume with cifs driver with same name and options as in point 2 - in plugin log now we have no Create call
7. Use volume with new container - tries to mount volume without my options - no success
8. `docker volume ls` still shows no volumes for cifs driver
9. `docker volume rm name` says that we have no such volume

So we can't create or delete such volume without restarting docker service.

**Comment by thaJeztah** [comment]
@IIIEII can you open a separate issue, because "volumes" are not related to the issue at hand here

**Comment by IIIEII** [comment]
@thaJeztah ok, I created new one #22155

**Comment by tnucera** [comment]
With Docker 1.11 : no issue

**Comment by Zapadlo** [comment]
We are also seeing this issue running CoreOS on AWS.

Docker version:
```
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.4
 Git commit:   3cd164c
 Built:
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.4
 Git commit:   3cd164c
 Built:
 OS/Arch:      linux/amd64
```

Docker info:
```
Containers: 38
 Running: 38
 Paused: 0
 Stopped: 0
Images: 31
Server Version: 1.10.3
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: none
Plugins:
 Volume: local
 Network: null host bridge
Kernel Version: 4.5.2-coreos
Operating System: CoreOS 1032.0.0 (MoreOS)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 15.67 GiB
Name: ip-172-24-121-215.eu-west-1.compute.internal
ID: S5TQ:B4NI:W7DL:57KS:TUZM:ZLKE:Y4HE:CR5E:RDQR:7AYA:IGMB:ZJH4
```

**Comment by antage** [comment]
I have this issue with docker 1.11.1.

**Comment by antage** [comment]
More detailed story:

Today I upgraded kernel to 4.5.0 and after reboot docker started to complain about 'Handler for POST /v1.23/containers/create returned error: Conflict. The name "[here is name of any container]" is already in use by container [here is uuid]. You have to remove (or rename) that container to be able to reuse that name."
I did can't start any container due to this error.
I tried to restart docker, uninstall docker and rm -rf /var/lib/docker and then install again. Nothing helps.
Only downgrading to kernel 4.4.0 solves the problem.

**Comment by anttiviljami** [comment]
:+1: started experiencing this issue after upgrading to 1.11.1

**Comment by anttiviljami** [comment]
Restarting the docker daemon in our case doesn't seem to help. Currently, we can't start a container by the same name as has been previously running. Any advice?

**Comment by anttiviljami** [comment]
```
# docker info
Containers: 135
 Running: 131
 Paused: 0
 Stopped: 4
Images: 652
Server Version: 1.11.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 723
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: null host bridge
Kernel Version: 3.13.0-86-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 10
Total Memory: 13.7 GiB
Name: -
ID: XOJO:TPPD:TXB4:7Z4Z:MH4L:LKFV:BMPG:3K4Z:ZPSF:Y7EU:JQMU:HW7P
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
```

**Comment by martingartonft** [comment]
Is there anything I can do or information I can provide to help diagnose this?

**Comment by justincormack** [comment]
@martingartonft can you provide a script that will reproduce it (eventually)?

**Comment by martingartonft** [comment]
I'll need to think about how to try and do that.
Currently it happens pretty infrequently.   Anecdotally, it seems to be when there are lots of containers being stopped and started around the same time (e.g., following a reboot) so might indicate some kind of race.
Let me see what I can manage.

**Comment by justincormack** [comment]
That seems plausible. The other option is to stare at the code a lot until you see the race condition, but that doesn't seem to have worked yet... But scripting a lot of containers and then doing demon restarts sounds like it might show it.

**Comment by pangzheng** [comment]
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 339
Server Version: 1.11.2
Storage Driver: overlay
 Backing Filesystem: xfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host bridge
Kernel Version: 3.10.0-327.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.703 GiB
Name: sry1
ID: URCV:RVOU:BKB2:FX37:CMHC:OPXR:IYXN:RXXW:HKAC:3ALE:XJ2I:ELF3
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled

**Comment by mhat** [comment]
We see the same issue, here's a script that appears to reproduce it on our systems.

```
while [ true ]; do
  echo "."
  docker run --read-only --rm  --name com.modeanalytics.test 1e6d584e7090 sleep 5  &
  sleep 0.1
  docker kill com.modeanalytics.test
  sleep 0.1
done
```

_You can use any image-id you like, the one above is a slightly modified debian image._

After a few seconds I get: 
```
docker: Error response from daemon: Conflict. The name "/com.modeanalytics.test" is already in use by container 5a3838613b8be7777d2c3311948c2a11b3eabe0a56fa88442cecf64f7e6cda38. You have to remove (or rename) that container to be able to reuse that name..
See 'docker run --help'.
Failed to kill container (com.modeanalytics.test): Error response from daemon: Cannot kill container com.modeanalytics.test: No such container: com.modeanalytics.test
```

#### docker --version
```
docker --version
Docker version 1.10.3, build 20f81dd
```

#### docker ps -all
```
docker ps --all | grep com.modeanalytics.test 
(empty)
```

#### docker rm 
```
docker rm com.modeanalytics.test
Failed to remove container (com.modeanalytics.test): Error response from daemon: No such container: com.modeanalytics.test
```

**Comment by thaJeztah** [comment]
@mhat I don't think that case would be solvable; there's no way to guarantee that the container is "gone" the moment it's killed; `docker run --rm` is a client-side action; the moment it detects that the connection with the container is lost, it performs a `docker rm`. If you try to start a new container with the same name *directly* after killing it, it may be either that `docker run rm` didn't detect yet that the connection was gone (because the container is killed), or the `docker rm` still being in progress

**Comment by mhat** [comment]
@thaJeztah I'm not so much worried about _the moment_ the container is killed. It's that there's a race where once the name is unavailable it will *never* be available again until the Docker daemon is restarted.

**Comment by thaJeztah** [comment]
@mhat but your example only shows the potentially "expected" failures (i.e. a race inside the loop), not a situation where a "ghost" container blocks access. FWIW, I tried running the script on docker 1.11.2, and so far haven't been able to reproduce the issue 😞

**Comment by mhat** [comment]
@thaJeztah Oh interesting. It's been hours and that name is still not available on the system I ran it on.

**Comment by thaJeztah** [comment]
@mhat no problem, thanks for the clarification. It's perfectly possible there's a race *somewhere*, these kind of bugs are really nasty to reproduce / test / resolve

**Comment by mhat** [comment]
@thaJeztah Nod. In production we see a container name become become unavailable one to two times a day across about a hundred containers. It's not frequent enough to pose a real issue but it's tedious. I'll see if I can get the steps to repro to work more reliable between 1.10 and 1.12 (the rels I have handy).

**Comment by cpuguy83** [comment]
This issue here is non-obvious other than that name generation/reservation is extremely convoluted and needs refactoring.

**Comment by cpuguy83** [comment]
If someone wants to take a stab at that, it would be super helpful.
Most of this is in daemon/names.go

**Comment by do3cc** [comment]
This might be the same issue as in #24706 
Our build system triggers this problem multiple times a week. If you can provide us tips on how to provide additional information, we would love to add more information.
So far we can see the container id that claims the name but cannot find the name with neither `ps -a` nor `docker inspect $ID`

**Comment by yamaszone** [comment]
+1

**Comment by bkleef** [comment]
I've updated from Docker 1.11.2 to 1.12.1. Will keep you guys posted!

**Comment by Umkus** [comment]
This just bit me as well on `1.13.1`.
* `docker rm -f name` gave `Error response from daemon: No such container: name`
* `docker ps -a` gave me an empty list
* `docker create --name name` gave `Error response from daemon: Conflict. The container name "/name" is already in use by container hash`

And those were sequential commands in systemd unit file.

**Comment by omegacoleman** [comment]
just issued this problem with systemd-nspawn!

I was running an arm container with qemu-user-static, and trying to do some disk io to a mounted loop device, suddenly the it got stuck.

I then closed the terminal, and then the problem happens. machinectl not work. 

If I'm right, this is because in Linux nobody can kill a process stucked in io problem. No matter you're what user (even root). I googled for the answer is that the SIGKILL sent to the process would be blocked by the process if it got into io problems, as signals for are processed async.

Technically it's not a bug for kernel -- it's the DESIGN eh?

So maybe, this is a dead-end -- to the container developers. Until kernel developers develop sth like SIGFXXK that would let u get rid of the process immediatly. 

I got my hands filled with matters now, back when I'm free I'll try repro it.

**Comment by kimattree** [comment]
Was having the same issue, although might be slightly unrelated...

In /usr/lib/systemd/system/docker.service

I removed "Mountflags=Slave" that a junior staff member had inserted from back when we used Docker 1.10 (now 1.13.1)

systemctl deamon-reload and docker restart and everything started working again.

**Comment by thaJeztah** [comment]
Let me close this ticket for now, as it looks like it went stale.
